CRASM- 3284: Develop automated Playwright tests to validate the filter persistence functionality implemented in CRASM-3004 for the VS Dashboard

jsalinasnttdata opened the pull request:
<!-- GitHub renders PRs such that soft line breaks are treated as hard
line breaks.  In order to make this template render as expected we
therefore have to avoid soft breaks and therefore will offend
markdownlint with long lines.  This is the reason for the markdownline
disable directive just below.

For more details see:
https://github.com/github/markup/issues/1050#issuecomment-294654762 -->
<!-- markdownlint-disable MD013 -->
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

This PR adds comprehensive automated Playwright tests to validate VS Dashboard filter persistence functionality across all user permission levels. The tests ensure that filter state is properly maintained during drill-down navigation and correctly reset during general navigation, while also verifying that user permissions are properly enforced.

**New Test Files Added:**
- `playwright/e2e/global-admin/vs-dashboard-filter-persistence.spec.ts` - 3 tests for full filter access
- `playwright/e2e/global-view/vs-dashboard-filter-persistence.spec.ts` - 3 tests for read-only filter access
- `playwright/e2e/regional-admin/vs-dashboard-filter-persistence.spec.ts` - 3 tests for scoped filter access
- `playwright/e2e/standard-user/vs-dashboard-filter-persistence.spec.ts` - 2 tests for disabled filter validation

**Test Coverage:**
- Filter persistence during drill-down navigation (Key Metrics, Domains tab)
- Filter reset during general navigation
- Browser back/forward navigation handling
- User permission enforcement validation
- Cross-user consistency verification

## 💭 Motivation and context ##

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

**Resolves CRASM-3284** - Develop automated Playwright tests to validate the filter persistence functionality implemented in CRASM-3004 for the VS Dashboard.

**Problem:** The VS Dashboard filter persistence feature (CRASM-3004) needed comprehensive automated testing to ensure it works correctly across all user permission levels and navigation scenarios. Without proper test coverage, there was risk of regression bugs in filter behavior and user permission enforcement.

**Solution:** Created a complete test suite that validates:
1. **Filter Persistence Logic** - Filters maintain state during drill-down navigation but reset during general page navigation
2. **User Permission Enforcement** - Different user types have appropriate filter access (Standard Users have filters disabled)
3. **Navigation Context Handling** - Proper distinction between drill-down vs general navigation
4. **Cross-Browser Compatibility** - Consistent behavior across different navigation patterns

The tests discovered that the original implementation was working correctly, and initial test failures were due to incorrect assumptions about expected behavior rather than feature bugs.

## 🧪 Testing ##

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

**Test Commands:**
```bash
# Run all filter persistence tests
npx playwright test e2e/*/vs-dashboard-filter-persistence.spec.ts

# Run specific user type tests
npx playwright test e2e/global-admin/vs-dashboard-filter-persistence.spec.ts  # 3/3 passing
npx playwright test e2e/global-view/vs-dashboard-filter-persistence.spec.ts   # 3/3 passing
npx playwright test e2e/regional-admin/vs-dashboard-filter-persistence.spec.ts # 3/3 passing
npx playwright test e2e/standard-user/vs-dashboard-filter-persistence.spec.ts  # 2/4 passing
```

# Run Report
```
npx playwright show-report playwright-report/html
```
**Global-Admin User:** 
<img width="1014" height="253" alt="Screenshot 2025-11-20 at 7 52 37 PM" src="https://github.com/user-attachments/assets/7bebedc5-bb31-48cb-8f6b-866fb6b57004" />

**Global-View User:**
<img width="1001" height="244" alt="Screenshot 2025-11-20 at 7 55 44 PM" src="https://github.com/user-attachments/assets/120ce631-f7b4-4e86-8d56-59bbea711538" />

**Regional-Admin User:** 
<img width="1011" height="256" alt="Screenshot 2025-11-20 at 8 00 07 PM" src="https://github.com/user-attachments/assets/37cd746b-d02b-45e3-89d0-b6dc87a0a446" />

**Standard User:**
<img width="1015" height="190" alt="Screenshot 2025-11-20 at 8 01 51 PM" src="https://github.com/user-attachments/assets/63c358f3-fc25-4fdc-b764-1e31ead78ab3" />

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] *All* future TODOs are captured in issues, which are referenced in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.
- [ ] Bump major, minor, patch, pre-release, and/or build versions [as appropriate](https://semver.org/#semantic-versioning-specification-semver) via the `bump_version` script *if* this repository is versioned *and* the changes in this PR [warrant a version bump](https://semver.org/#what-should-i-do-if-i-update-my-own-dependencies-without-changing-the-public-api).
- [ ] Create a pre-release (necessary if and only if the pre-release version was bumped).

## ✅ Pre-merge checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- These boxes should remain unchecked until the pull request has been -->
<!-- approved. -->

- [ ] Revert dependencies to default branches.
- [ ] Finalize version.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [ ] Create a release (necessary if and only if the version was bumped).

jsalinasnttdata committed:
refs CRASM-3284: Add comprehensive VS Dashboard filter persistence tests for all user types

- Add automated Playwright tests for VS Dashboard filter persistence across all user permission levels
- Implement tests for Global Admin, Global View, Regional Admin, and Standard User roles
- Validate filter persistence during drill-down navigation and reset during general navigation
- Verify user permission enforcement (Standard Users have filters disabled)

jsalinasnttdata committed:
Merge branch 'develop' into jsalinas-playwright-filter-persistence-tests-CRASM-3284

jsalinasnttdata committed:
refs CRASM-3284: Precommit hook

jsalinasnttdata committed:
refs CRASM-3284: Clean up

jsalinasnttdata committed:
Merge branch 'develop' into jsalinas-playwright-filter-persistence-tests-CRASM-3284

jsalinasnttdata committed:
Merge branch 'develop' into jsalinas-playwright-filter-persistence-tests-CRASM-3284

jsalinasnttdata committed:
refs CRASM-3284: Consolidate VS Dashboard filter helper functions

- Extract repeated filter helper functions from test files into shared utils/filters.ts
- Add openVSFiltersDrawer(), closeVSFiltersDrawer(), and checkFilterState() functions
- Create generic checkFilterState() that handles both value checking and disabled status

jsalinasnttdata committed:
Merge branch 'develop' into jsalinas-playwright-filter-persistence-tests-CRASM-3284

jsalinasnttdata committed:
refs CRASM-3284: Update comment

jsalinasnttdata committed:
refs CRASM-3284: Precommit hook

cduhn17 commented on playwright/e2e/global-view/vs-dashboard-filter-persistence.spec.ts:
```diff
@@ -0,0 +1,191 @@
+    // Step 3: Browser back navigation
+    console.log('Step 3: Using browser back button');
+    await page.goBack();
+    await page.waitForURL('**/VSDashboard**');
```
i see that in other places you are using ROUTES.VSDASHBOARD, should we be using this constant here?

cduhn17 commented on playwright/e2e/regional-admin/vs-dashboard-filter-persistence.spec.ts:
```diff
@@ -0,0 +1,221 @@
+
+test.describe('VS Dashboard Filter Persistence - Regional Admin User', () => {
+  // Drill-down via Key Metrics persists filters
+  test('Drill-down via Key Metrics persists filters', async ({
```
Consider centralizing this as you are using this in multiple locations.

jsalinasnttdata replied:
This was discussed in recent discussion with the test team. It was decided that repetition is fine for now as many of the permissions will be changing very soon.

cduhn17 commented on playwright/e2e/regional-admin/vs-dashboard-filter-persistence.spec.ts:
```diff
@@ -0,0 +1,221 @@
+
+    // Try to select from available regions
+    try {
+      await selectFromAutocomplete(page, /^region$/i, /Region\s*[1-4]/i);
```
Consider adding an assertion that the region input has some value or is disabled, so you still assert something  about permissions.

cduhn17 commented on playwright/e2e/standard-user/vs-dashboard-filter-persistence.spec.ts:
```diff
@@ -0,0 +1,67 @@
+
+    // Step 1: Verify page loads successfully
+    console.log('Step 1: Verifying VS Dashboard loads for Standard User');
+    await expect(page).toHaveURL(/.*VSDashboard.*/);
```
Can we use the constant `ROUTES.VSDASHBOARD` here?

cduhn17 commented on playwright/utils/filters.ts:
```diff
@@ -366,3 +366,124 @@ export async function urlHasBothFilters(
+        };
+      }
+    } catch (error) {
+      console.log('Error checking filters - assuming filters not available');
```
Consider logging the error as well to help CI triage. `console.log('Error checking filters - assuming filters not available', error);` In some instances we will have to be careful when logging `error` so we dont leak sensitive data.

cduhn17 commented on playwright/utils/filters.ts:
```diff
@@ -366,3 +366,124 @@ export async function urlHasBothFilters(
+          org: currentOrgValue,
+          hasFilters:
+            currentRegionValue !== '' &&
+            !currentRegionValue.includes('All Regions') &&
```
This is fine, just keep in mind that it is hard coded to the exact text "All Regions". If that content is localized or adjusted, you will need to update this logic. You might eventually want to hook this into the same config that defines the options for the region filter.

cduhn17 reviewed (commented):
See comments. Nothing major, just some "house keeping".If you have questions or concerns let me know.

jsalinasnttdata committed:
Merge branch 'develop' into jsalinas-playwright-filter-persistence-tests-CRASM-3284

jsalinasnttdata committed:
refs CRASM-3284: Replace hardcoded routes with ROUTES constants in filter persistence tests

- Replace '**/inventory/vulnerabilities**' with `**${ROUTES.VULNERABILITIES}**`
- Replace '**/VSDashboard**' with `**${ROUTES.VSDASHBOARD}**`

jsalinasnttdata committed:
refs CRASM-3284: Replace hardcoded routes with ROUTES constants in filter persistence tests for standard user

jsalinasnttdata committed:
refs CRASM-3284: Add error logging to checkFilterState for CI triage

jsalinasnttdata committed:
refs CRASM-3284: Align allRegionsOption constant with component pattern

- Follow the same pattern as VSDashRegionAndOrgFilters.tsx

jsalinasnttdata committed:
refs CRASM-3284: Add permission assertion for regional admin region filter

jsalinasnttdata committed:
refs CRASM-3284: Update regional admin tests to reflect current filter permissions

- Regional admin users now have the same filter capabilities as global admin and global view users.

jsalinasnttdata committed:
refs CRASM-3284: Precommit hook

cduhn17 reviewed (approved):
LGTM
